dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java

```diff
@@ -0,0 +1,270 @@
+                String includedString = includedAppList.get().get(0).toLowerCase().trim();
+                if (includedString.equals("all")) {
+                    configMode = Optional.of(ConfigMode.All);
+                    includedAppList = Optional.empty();
+                } else if (includedString.equals("first")) {
+                    configMode = Optional.of(ConfigMode.First);
+                    includedAppList = Optional.empty();
+                } else if (includedString.equals("none")) {
+                    configMode = Optional.of(ConfigMode.None);
+                    includedAppList = Optional.empty();
+                }
+            }
+
+            //store outside of an optional because merging two optional lists is convoluted
+            includedApps.addAll(includedAppList.orElse(Collections.emptyList()));
+            excludedApps.addAll(excludedAppList.orElse(Collections.emptyList()));
+            includedModules.addAll(includedModulesList.orElse(Collections.emptyList()));
+            excludedModules.addAll(excludedModulesList.orElse(Collections.emptyList()));
+
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "OpenAPI finished processing modules from server.xml."
+                             + " found the following configuration for included apps {0},"
+                             + " found the following configuration for excluded apps {1},"
+                             + " found the following configuration for included modules {2},"
+                             + " found the following configuration for excluded modules {3}",
+                         includedApps, excludedApps, includedModules, excludedModules);
+            }
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (other == null) {
+                return false;
+            }
+            if (other == this) {
+                return false;
+            }
+            if (!(other instanceof MpConfigServerConfigObject)) {
+                return false;
+            }
+
+            MpConfigServerConfigObject otherConfig = (MpConfigServerConfigObject) other;
+
+            //We sort all the lists here because we only need ordering for equality
+            //We only check equality if the server.xml is updated while the app is live
+            //And we don't want to pay the performance cost unless that happens.
+
+            includedApps.sort(Comparator.naturalOrder());
+            excludedApps.sort(Comparator.naturalOrder());
+            includedModules.sort(Comparator.naturalOrder());
+            excludedModules.sort(Comparator.naturalOrder());
+
+            otherConfig.includedApps.sort(Comparator.naturalOrder());
+            otherConfig.excludedApps.sort(Comparator.naturalOrder());
+            otherConfig.includedModules.sort(Comparator.naturalOrder());
+            otherConfig.excludedModules.sort(Comparator.naturalOrder());
```

Comment:
We mustn't modify the object's state in the equals method. This could cause a concurrent modification exception if another thread is reading these fields while the sort is occurring.

We could either:
- Not sort before doing the comparison
- Copy before sorting
- Copy into a set instead of sorting
- Store the data in a set rather than a list in the first place

I'm not sure which is the best option. I'm not sure that we really need to sort before comparing since if the config is not changed in the server.xml, we should get the elements back in the same order.

On the other hand, if we really don't care about the order, storing the data in sets instead of lists makes more sense.

Reply:
I was thinking of one edge case where the before is 

```
<includeApplication>one</includeApplication>
<includeApplication>two</includeApplication>
```

And the after is 

```
<includeApplication>two</includeApplication>
<includeApplication>one</includeApplication>
```

Which is logically equivalent but maybe that's not worth the trouble

Reply:
The reason its not a set was thinking about someone doing 
```
<includeApplication>one</includeApplication>
<includeApplication>one</includeApplication>
```
I thought it was best to accurately represent that info inside java, encase we want to use it for something.